gnomad_constraint/utils/constraint_basics.py

```diff
@@ -79,3 +86,79 @@ def prepare_ht_for_constraint_calculations(ht: hl.Table):
+) -> hl.Table:
+    """
+    Count the observed variants and possible variants by substitution, context, methylation level, and additional `grouping`.
+
+    Variants where there was no coverage, a low-quality variant was observed, or a variant
+    above 0.1% frequency was observed  in `exome_ht` and `context_ht` was removed. For each substitution,
+    context, and methylation level, observed variants were counted using `exome_ht`, and possible variants
+    was counted using `context_ht` that anti joins `exome_ht`.
+
+    The returned Table includes following annotations:
+        - context - trinucleotide genomic context
+        - ref - the middle base of `context`
+        - alt - the alternate base
+        - methylation_level - methylation_level
+        - observed_variants - observed variant counts in `exome_ht`
+        - possible_variants - possible variant counts in `context_ht`
+        - downsampling_counts_{pop} - variant counts in downsaplings for populations in `pops`
+        - mu_snp - SNP mutation rate
+        - annotations added by `annotate_mutation_type`
+
+    :param exome_ht: Preprocessed exome Table.
+    :param context_ht: Preprocessed context Table.
+    :param mutation_ht: Preprocessed mutation rate Table.
+    :param max_af: Maximum allele frequency for a variant to be included in returned counts. Default is 0.001.
+    :param keep_annotations: Annotations to keep in the context Table.
+    :param pops: List of populations for choosing downsamplings when counting variants. Default is ().
+    :param grouping: Annotations other than 'context', 'ref', 'alt' to group by when counting variants. Default is ('exome_coverage',).
+    :param partition_hint: Target number of partitions for aggregation. Default is 100.
+    :return: Table with observed variant and possible variant count.
+    """
+    # It's adjusted allele frequency information for all release samples in gnomAD
+    freq_expr = exome_ht.freq[0]
+
+    keep_criteria = (
+        (freq_expr.AC > 0) & exome_ht.pass_filters & (freq_expr.AF <= max_af)
+    )
+
+    ht = count_variants_by_group(
+        filter_vep_transcript_csqs(exome_ht.filter(keep_criteria)).select(
+            *list(keep_annotations) + ["freq"]
+        ),
+        additional_grouping=grouping,
+        partition_hint=partition_hint,
+        count_downsamplings=pops,
+        use_table_group_by=True,
+    )
+    ht = ht.transmute(observed_variants=ht.variant_count)
+
+    context_ht = filter_vep_transcript_csqs(context_ht).select(*keep_annotations)
+    context_ht = context_ht.anti_join(exome_ht.filter(keep_criteria, keep=False))
+    possible_ht = count_variants_by_group(
+        context_ht, additional_grouping=grouping, use_table_group_by=True
+    )
+    possible_ht = annotate_with_mu(possible_ht, mutation_ht)
+    possible_ht = possible_ht.transmute(possible_variants=possible_ht.variant_count)
+
+    ht = ht.join(possible_ht, "outer")
+    ht = annotate_mutation_type(ht)
```

Comment:
We try to add our parameters that we use as globals in many cases so that users know what parameters were used to create the Table. If you think it doesn't make sense in this case that is fine, but here are some examples if helpful to look at:
https://github.com/broadinstitute/gnomad_methods/blob/11940136e64ef3a1d5329c2a5f88b3ad317b8d1a/gnomad/sample_qc/pipeline.py#L224
https://github.com/broadinstitute/gnomad_methods/blob/11940136e64ef3a1d5329c2a5f88b3ad317b8d1a/gnomad/sample_qc/pipeline.py#L339
https://github.com/broadinstitute/gnomad_qc/blob/155bbdad2d3cb05d638bc07b4365f5f0ea8a3e85/gnomad_qc/v4/sample_qc/hard_filters.py#L108
https://github.com/broadinstitute/gnomad_qc/blob/155bbdad2d3cb05d638bc07b4365f5f0ea8a3e85/gnomad_qc/v3/sample_qc/sample_qc.py#L704
https://github.com/broadinstitute/regional_missense_constraint/blob/65d46f3988e2dbed617ad27aa780d8f0fc558a5c/rmc/pipeline/regional_constraint.py#L123
https://github.com/broadinstitute/regional_missense_constraint/blob/65d46f3988e2dbed617ad27aa780d8f0fc558a5c/rmc/pipeline/regional_constraint.py#L304